### ISSUE TITLE
Versioning needed

### ISSUE THREAD

**Issue by gregmagolan** [bazelbuild/bazel-watcher#77]
It would be good to get some some version info so you could type `ibazel --version` and see what you're running. There doesn't seem to be a standard way to version go project however.

Where would the version live? Should it be sematic versioning? Right now there is an version number npm/package.json. Should this be the source of truth for the version number? If so, how to consume that number and get it into the go binary?

How would releases work? Update version number, tag commit and publish to npm? Noticed that there is a version 0.0.1 published to npm.

**Comment by gregmagolan** [comment]
On a related note, could also use a check for a minimum compatible bazel version

**Comment by alexeagle** [comment]
Note, Angular has a similar need. The `workspace_status_command` flag to bazel lets you inject a script that reads the current SHA of the git repo, but then you also need to inject that information into the binary.
https://github.com/bazelbuild/bazel/issues/2898 does not look promising

**Comment by achew22** [comment]
WRT doing it in golang it is trivial using [`go_binary`](https://github.com/bazelbuild/rules_go/blob/master/go/core.rst#go_binary)'s x_defs command.

```
...
x_defs = {
    "github.com/bazelbuild/bazel-watcher/package.GIT_VERSION": "{STABLE_GIT_VERSION}",
}
...
```

Workplace status:
```
#! /usr/bin/env bash

echo -n "STABLE_GIT_VERSION "
if git status --porcelain > /dev/null 2>&1; then
  git --no-pager log -1 --format="%h"
else
  git --no-pager log -1 --format="%h-dirty"
fi
```
and an invocation with `--workspace_status_command=path/to/workplace_status.sh`

If you're working in a rules_* that doesn't have the nice stamping stuff of rules_go, you can use a genrule to  get the same results:

```
genrule(
    name = "build_ctx",
    outs = ["build_ctx.ts"],
    cmd = """
# Render a header
cat > $(location :build_ctx.ts) <<EOF
// @filedescription DO NOT EDIT. THIS FILE IS AUTOMATICALLY GENERATED.

// Some prefix or preamble about stuff and things or maybe the header for an Angular module.
EOF
# Compile in the build information
for file in bazel-out/stable-status.txt bazel-out/volatile-status.txt; do
  # Put a newline in to prevent accidentally putting 2 consts on one line.
  echo "\n// From $$file" >> $(location :build_ctx.ts)
  cat $$file | sed "s|\(STABLE_\)\?\(.*\) \(.*\)|const '\\2' = '\\3';|" >> $(location :build_ctx.ts)
done
# Now a footer
cat >> $(location :build_ctx.ts) <<EOF
// Some kind of export statement or footer

// DO NOT EDIT. THIS FILE IS AUTOMATICALLY GENERATED.
EOF
""",
    stamp = 1,
)
```

**Comment by gregmagolan** [comment]
Does relying on a `workplace_status.sh` shell script mean that it won't on Windows?

**Comment by achew22** [comment]
If there is no value it will not replace the string. That said, it shouldn't matter since the binaries will get built on a CI box that is almost guaranteed to be Linux.

**Comment by gregmagolan** [comment]
Put up PR #79 to parse the version out of package.json and included it in the binary

**Comment by dougkoch** [comment]
Merged #79, but leaving this issue open until there is a consensus on what to do about enforcing a minimum bazel version.

**Comment by achew22** [comment]
There are two potential version problems.

1.  Problems with dependencies required to build iBazel. These are totally standard for a project and fairly well protected against by explicitly stating all of our dependencies in the `WORKSPACE`file. The build should even be reproducible as a cross-compile from linux and darwin to all target platforms. I'm willing to consider this effectively solved.
1.  Problems with the dependent version of Bazel. Right now we are using the `STDOUT`/CLI interface for interacting with Bazel. Any changes in that system would require significant forewarning breaking a lot of internal to Google tooling. As a structural matter, I think those APIs are essentially locked in stone. That said, I would be remiss to not quote the SRE adage, "hope isn't a strategy." With #82 merged we are now testing against regression against old versions of Bazel (only back to 0.7.0 right now). As new Bazels are released, we can add to the set of Bazels we build against (including RCs).

As such, I'm going to close this issue considering it solved. If more problems pop up, let's open a new issue describing the specifics.
